core/jobs/io/gcs_io.py

```diff
@@ -0,0 +1,116 @@
+# cannot subclass 'PTransform' (has type 'Any')), we added an ignore here.
+class ReadFile(beam.PTransform): # type: ignore[misc]
+    """Read files form the GCS."""
+
+    def __init__(
+        self,
+        client: Optional[gcsio_test.FakeGcsClient] = None,
+        mode: str = 'r',
+        label: Optional[str] = None
+    ) -> None:
+        super().__init__(label=label)
+        self.client = client
+        self.mode = mode
+
+    def expand(
+       self, filenames: beam.PCollection
+    ) -> beam.PCollection[datastore_services.Model]:
+        """Returns PCollection with file data."""
+        return (
+            filenames
+            | 'Read the file' >> beam.Map(self._read_file)
+        )
+
+    def _read_file(self, filename: str) -> bytes:
+        """Helper function to read the contents of a file."""
+        gcs = gcsio.GcsIO(self.client)
+        return gcs.open(filename, mode=self.mode).read()
+
+
+# TODO(#15613): Here we use MyPy ignore because of the incomplete typing of
+# apache_beam library and absences of stubs in Typeshed, forces MyPy to
+# assume that PTransform class is of type Any. Thus to avoid MyPy's error (Class
+# cannot subclass 'PTransform' (has type 'Any')), we added an ignore here.
+class WriteFile(beam.PTransform): # type: ignore[misc]
+    """Write files to GCS."""
+
+    def __init__(
+        self,
+        client: Optional[gcsio_test.FakeGcsClient] = None,
+        mode: str = 'w',
+        mime_type: str = 'application/octet-stream',
+        label: Optional[str] = None
+    ):
+        super().__init__(label=label)
+        self.client = client
+        self.mode = mode
+        self.mime_type = mime_type
+
+    def expand(
+        self, filenames: beam.PCollection) -> beam.PCollection:
+        """Returns the PCollection of files that have written to the GCS."""
+        return (
+            filenames
+            | 'Write files to GCS' >> beam.Map(self._write_file)
+        )
+
+    def _write_file(self, filename):
+        """Helper function to write file to the GCS."""
+        gcs = gcsio.GcsIO(self.client)
+        bucket, _ = gcsio.parse_gcs_path(filename['file'])
+        print("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")
+        print(gcs.get_bucket(bucket))
+        if gcs.get_bucket(bucket + 'abc') is None:
+            # gcs.create_bucket(bucket)
+            print("bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb")
+        if not gcs.exists(filename['file']):
+            file = gcs.open(
+                filename=filename['file'],
+                mode=self.mode,
+                mime_type=self.mime_type)
+            write_file = file.write(filename['data'])
+            file.close()
+            print("******************************")
+            print(self.client.buckets)
+            print(self.client.buckets.get_bucket('gcsio-test'))
+            print(gcs.get_bucket(bucket))
+            print(self.client.objects.files)
+            return write_file
```

Comment:
Hey @vojtechjelinek,
I have added these print statements to debug the issue we are facing, I will post the results below. I think the `gcsio` library that we are using is not working correctly, please note that on line `101` I check if the gcs has a bucket or not, I'm checking a fake bucket but in the below results you will not see anywhere printing `bbbbbbbbbb...`. The `get_bucket` should return None but it is not. Please note that on line I try to print `gcsio-test` bucket if it exists or not, in the result below you will be able to see it, If I try to print some random bucket it will print that too.
**Result**
```
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
<Bucket
 acl: []
 cors: []
 defaultObjectAcl: []
 name: 'gcsio-test'
 projectNumber: 1>
******************************
<apache_beam.io.gcp.gcsio_test.FakeGcsBuckets object at 0x7f0d88112160>
<Bucket
 acl: []
 cors: []
 defaultObjectAcl: []
 name: 'gcsio-test'
 projectNumber: 1>
<Bucket
 acl: []
 cors: []
 defaultObjectAcl: []
 name: 'gcsio-test'
 projectNumber: 1>
{('gcsio-test', 'dummy_file_1'): <apache_beam.io.gcp.gcsio_test.FakeFile object at 0x7f0d88124520>}
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
<Bucket
 acl: []
 cors: []
 defaultObjectAcl: []
 name: 'gcsio-test'
 projectNumber: 1>
******************************
<apache_beam.io.gcp.gcsio_test.FakeGcsBuckets object at 0x7f0d88112160>
<Bucket
 acl: []
 cors: []
 defaultObjectAcl: []
 name: 'gcsio-test'
 projectNumber: 1>
<Bucket
 acl: []
 cors: []
 defaultObjectAcl: []
 name: 'gcsio-test'
 projectNumber: 1>
{('gcsio-test', 'dummy_file_1'): <apache_beam.io.gcp.gcsio_test.FakeFile object at 0x7f0d88124520>, ('gcsio-test', 'dummy_file_2'): <apache_beam.io.gcp.gcsio_test.FakeFile object at 0x7f0d881246a0>}
```
Thanks

Reply:
Resolving this as it is happening due to `FakeGcsBuckets`.
Thanks